Skatepark: add the header

jffng opened the pull request:
<!-- Thanks for contributing to our free themes! Please provide as much information as possible with your Pull Request by filling out the following - this helps make reviewing much quicker! -->

#### Changes proposed in this Pull Request:

This PR adds a header template part and styles per the design for Skatepark. It applies some ideas from the every-layout.dev book. It was pretty interesting to explore and grok the concept of truly flexible layouts not tied to breakpoints. 

Looking at Skatepark's header design and the components involved, there are three states to consider:

A. Not wrapped | B. Wrapped | C. Wrapped w/ responsive navigation 
------ | ------ | ------
<img width="1039" alt="Screen Shot 2021-08-03 at 12 32 10 PM" src="https://user-images.githubusercontent.com/5375500/128053361-2336062c-3587-42b6-9ea1-0ac078c31ee6.png"> | <img width="692" alt="Screen Shot 2021-08-03 at 12 32 03 PM" src="https://user-images.githubusercontent.com/5375500/128053376-79e9f043-6f91-490d-bfde-a4182a9c0974.png"> | <img width="642" alt="Screen Shot 2021-08-03 at 12 31 54 PM" src="https://user-images.githubusercontent.com/5375500/128053401-6b147bfa-21d1-4d8c-bb18-8eb44cadf5bf.png">

B is the interesting scenario. Before, I would have tried to write a media query that tells the header to break onto two lines at a specific pixel value. But this is not a great approach because we don't know for certain the content width / number of menu items, thus making it impossible to write a media query that solves for all the situations. 

I could not get around using a media query for the navigation + social links, which I think is acceptable because the navigation block's responsive behavior is tied to the same media query.

The spacing needs some tightening to match the comps, but I wanted to get your feedback in the meantime. 

#### Related issue(s):

#4313

jffng committed:
Create the scaffolding for header template part.

jffng committed:
Use intrinsic design layouts for the primary header container.

jffng committed:
Simplify flex wrapping rules.

jffng committed:
Comments for everylayout.dev pages.

jffng commented on skatepark/block-template-parts/header.html:
```diff
@@ -0,0 +1,25 @@
+
+<!-- wp:group -->
+<div class="wp-block-group nav-links"><!-- wp:navigation {"orientation":"horizontal","itemsJustification":"right","isResponsive":true,"style":{"typography":{"fontStyle":"normal","fontWeight":"900","textTransform":"uppercase"}},"fontSize":"small"} -->
+<!-- wp:navigation-link {"label":"Home","type":"custom","url":"http://skateparkdev.local/","kind":"custom","isTopLevelLink":true} /-->
```
Not suggesting we ship the theme with these links baked in, but I included them so it's maybe easier to test.

jffng committed:
Use the Gutenberg flex layout.

MaggieCabrera commented on skatepark/sass/templates/_header.scss:
```diff
@@ -0,0 +1,33 @@
+header {
+	justify-content: space-between; // Apply a cluster (flex?) layout
+	flex-wrap: wrap-reverse;
```
TIL!

MaggieCabrera commented on skatepark/sass/templates/_header.scss:
```diff
@@ -0,0 +1,33 @@
+			align-items: center;
+
+			&::after {
+				content: "Menu";
```
We can't translate this :(

mikachan replied:
Could we use the lang attribute on the `html` tag to translate this? Something like:

```
html[lang="en-US"] {
	.nav-links {
		.wp-block-navigation__responsive-container-open {
			&::after {
				content: "Menu";
			}
		}
	}
}
```

I guess this would need to be added via a PHP template in an inline style.. Or is this horrible and hacky?

MaggieCabrera replied:
we could add this line of CSS as inline inside functions.php and wrap the string around the PHP translation function directly, no need to target the lang attribute. It's not much better, but it's another option

mikachan replied:
Ah nice ok! Well, we have options 😄

jffng replied:
Ideally there is an option in the nav block itself to provide this label next to the responsive hamburger. I wonder if it's planned

MaggieCabrera commented:
I love how this looks! And it's very simple, great job!

This is what I see:

<img width="298" alt="Screenshot 2021-08-04 at 10 09 47" src="https://user-images.githubusercontent.com/3593343/128145930-225b16bb-db6b-4586-9b18-8839522c6c52.png">

The social icons could have the same treatment as the footer block so that they align with the menu. In this particular case, I'd add a class to the block, since the users may want to re build the header or use another theme's and we don't want the css to break that.

This is what the footer is doing:
```
	.wp-block-social-links.is-style-logos-only {
		margin-left: calc( -1 * ( 8px + 0.25em ) );
	}
```

If either the navigation or the social links are not present, this happens. The scenario with no social links is worth considering.

<img width="1089" alt="Screenshot 2021-08-04 at 10 13 22" src="https://user-images.githubusercontent.com/3593343/128146435-401e8aac-626a-414b-a3bd-62c1d2b4f20f.png">

<img width="1086" alt="Screenshot 2021-08-04 at 10 15 13" src="https://user-images.githubusercontent.com/3593343/128146734-0bc02f5b-0204-4ff4-b6c3-7af5dfce63e6.png">


And right now Skatepark doesn't support a site logo, should we add support for it, @melchoyce If so, should we copy Quadrat's behavior for it?

mikachan commented on skatepark/sass/templates/_header.scss:
```diff
@@ -0,0 +1,33 @@
+	flex-wrap: wrap-reverse;
+	> * { 
+		flex-grow: 1;
+		> * { // Apply a stack layout (page 69 of the every-layout.dev PDF) 
+			margin-top: 20px;
+			margin-bottom: 20px;
+		}
```
This is cool, I love the idea of styling the context rather than the element itself.

mikachan reviewed (commented):
This looks ace! Agree with @MaggieCabrera's comments. Looking good in all browsers and resolutions. I agree with needing to use the media query for the navigation and social links, I can't think of another way we could do this.

melchoyce commented:
Whoo 🎉 

Minor notes:
- The last icon in the social block has a right margin. Can we remove that margin for the last item in the menu?
- The tagline's font-weight should be medium
- Navigation items are missing some letter spacing (`0.1em`)

> And right now Skatepark doesn't support a site logo, should we add support for it, @melchoyce If so, should we copy Quadrat's behavior for it?

Yes and yes, though I think I'll want to test once we have Quadrat's behavior in Skatepark.

jffng committed:
Refine header styles and translate the menu label.

jffng commented on skatepark/functions.php:
```diff
@@ -40,5 +40,13 @@ function skatepark_support() {
 	wp_enqueue_style( 'skatepark-styles', get_stylesheet_directory_uri() . '/assets/theme.css', array( 'blockbase-ponyfill' ), wp_get_theme()->get( 'Version' ) );
+
+	// Allow the responsive menu label to be translated.
+	$menu_label_css = '
```
This feels very messy, which makes me think it should go in the block.

In the meantime, maybe it should live in blockbase.

MaggieCabrera replied:
but blockbase doesn't have the Menu text. I think this is probably [planned for the navigation block](https://github.com/WordPress/gutenberg/issues/22824#issuecomment-789584886)

jffng replied:
Ok, I'm going to remove it from the theme for now and open an issue for it in Gutenberg.

jffng commented on skatepark/theme.json:
```diff
@@ -381,7 +381,8 @@
 				"typography": {
-					"fontSize": "var(--wp--preset--font-size--normal)"
+					"fontSize": "var(--wp--preset--font-size--normal)",
+					"letterSpacing": "0.05em"
```
@melchoyce I added the letter spacing for all navigation blocks — is that okay, or do you think it should be targeted to the navigation block in the header only?

jffng replied:
Also 0.1em for letter spacing looked like a lot compared to the comps, so I used 0.05 but let me know if you think it should be 0.1em.

jffng commented:
Thanks for the reviews and comments!

I made the following changes:

- Removed right margin from social links
- Changed the flex alignment to handle the case when either no menu or social links are present
- Made it so the menu label can be translated
- Added a placeholder for the site logo in the template. 

I think this is ready for another review, be sure to test out adding a logo.

jffng committed:
Add site logo.

MaggieCabrera commented on skatepark/block-template-parts/header.html:
```diff
@@ -0,0 +1,26 @@
+<!-- wp:navigation-link {"label":"Contact","type":"page","id":14,"url":"http://skateparkdev.local/contact/","kind":"post-type","isTopLevelLink":true} /-->
+<!-- /wp:navigation -->
+
+<!-- wp:social-links {"iconColor":"primary","iconColorValue":"#000000","className":"items-justified-right is-style-logos-only"} -->
```
we could use the primary value here so that it changes when we change palettes

jffng replied:
Oooo good idea.

MaggieCabrera commented:
The logo has a few differences comparing to Quadrat:

<img width="481" alt="Screenshot 2021-08-05 at 18 14 16" src="https://user-images.githubusercontent.com/3593343/128386099-3a0627e5-3c1f-4c8c-a050-f1969c502ba6.png">
<img width="1346" alt="Screenshot 2021-08-05 at 18 14 07" src="https://user-images.githubusercontent.com/3593343/128386102-31ae2f0e-7cc7-4ea2-9f2b-fbb79d473244.png">
<img width="815" alt="Screenshot 2021-08-05 at 18 13 07" src="https://user-images.githubusercontent.com/3593343/128386105-8664f579-b7c5-4f46-a7ce-3e57afbb0fa2.png">
<img width="507" alt="Screenshot 2021-08-05 at 18 13 00" src="https://user-images.githubusercontent.com/3593343/128386106-4825c965-8916-43ab-889a-ca647bb865f3.png">

melchoyce commented:
Can we swap out the menu icon for a slightly thicker, three-bar version? 

![image](https://user-images.githubusercontent.com/2846578/128392572-6d80066e-674e-4bf3-b632-a98ff17db62e.png)

And if possible, open the menu underneath instead of overlaying the header:

![image](https://user-images.githubusercontent.com/2846578/128392765-3575d079-11ba-4484-abd3-1f39bd58e97b.png)

If that's not possible, we can keep as-is — just feels especially annoying to open the menu on the left side and then close it on the right. Also happy with us playing with this later in a new PR.

jffng committed:
Use css var instead of baked color.

jffng commented:
> The logo has a few differences comparing to Quadrat:

@melchoyce any guidance on the position of the logo across viewports? The header on Quadrat is different so I wasn't sure how to handle. 

> And if possible, open the menu underneath instead of overlaying the header:

I think this will be tricky / I'm not sure if it's possible with the nav block currently. We can at least move the position of the close button within the modal so it's the same location as the open button.

melchoyce commented:
Would it work to display side-by-side on desktop, and stack on mobile?

![image](https://user-images.githubusercontent.com/2846578/128413593-2e869f7c-a704-4bfe-a825-522006c832b8.png)

MaggieCabrera commented:
Just adding that this also closes #4319 since I don't think we need a separate issue for that

jffng committed:
Move flex layout to group so we can reuse templates from blockbase.

jffng committed:
Comment why flex grow is needed.

jffng committed:
Use font-weight support in Gutenberg.

jffng committed:
Remove menu label in favor of adding it in Gutenberg.

jffng commented:
> Would it work to display side-by-side on desktop, and stack on mobile?

I gave this a try. Unless there is a purely flex based approach that I am missing, I don't think it's worth it because it would require another group block and more media-query based CSS, which is very fragile when we open things to the site editor.
 
> The tagline's font-weight should be medium

I opened a small change in Gutenberg to allow this so we don't have to add CSS for it and the user can customize it: https://github.com/WordPress/gutenberg/pull/33983

I also removed the responsive menu label in favor of adding the ability to the nav block itself: https://github.com/WordPress/gutenberg/issues/33985

Thoughts on bringing this in and iterating?

jffng committed:
Remove hard coded nav links.

mikachan commented:
This is looking good for me in all browsers and resolutions. I've tested with and without both a logo and a menu.

Bringing this in and iterating sounds good to me.

MaggieCabrera reviewed (approved):
+1 on bringing this in and iterating. And I also agree that the menu label is best handled on the editor.
